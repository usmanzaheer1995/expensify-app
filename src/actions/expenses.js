import uuid from 'uuid';

import database from '../firebase/firebase';

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

export const addExpense = (expense = {}) => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense,
        }));
      })
      .catch((err) => console.log(err));
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// async
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    return database.ref('expenses').once('value', (snapshot) => {
      const expenses = [];
      snapshot.forEach((snapshotItem) => {
        expenses.push({
          id: snapshotItem.key,
          ...snapshotItem.val()
        });
      });
      dispatch(setExpenses(expenses));
    })
  }
}