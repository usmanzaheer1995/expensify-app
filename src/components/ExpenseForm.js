import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? props.expense.amount.toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      edit: props.edit ? true : false,
      error: ''
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;

    // visit regex101.com and find out what this regex does: ^\d{1,}(\.\d{0,2}?)?$
    if (!amount || amount.match(/^\d{1,}?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount!'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10),
        createdAt: this.state.createdAt.valueOf(), // to get back timestamp instead of a moment object (in ms)
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            placeholder="Description"
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onDateChange} // PropTypes.func.isRequired
            focused={this.state.calendarFocused} // PropTypes.bool
            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
            numberOfMonths={1}
            isOutsideRange={day => false}
            displayFormat="DD/MM/YYYY"
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense (optional)"
          />
          <button>{this.state.edit ? 'Edit expense' : 'Add expense'}</button>
        </form>
      </div>
    );
  }
}
