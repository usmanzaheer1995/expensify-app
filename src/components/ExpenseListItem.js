import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

export const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {amount} - {moment(createdAt).format('DD/MM/YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;
