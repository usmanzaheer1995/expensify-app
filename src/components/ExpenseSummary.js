import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import numeral from "numeral";
import converter from "number-to-words";

import selectExpenses from "./../selectors/expenses";
import expensesTotal from "./../selectors/expenses-total";

export const ExpenseSummary = ({ expenses, expensesTotal }) => {
  const expenseWord = expenses.length === 1 ? "expense" : "expenses";
  // if (expenses.length === 0) return <div />;
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{converter.toWords(expenses.length)}</span> {expenseWord} totaling{" "}
          <span>{"Rs " + numeral(expensesTotal).format("0,0")}</span>.
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    expensesTotal: expensesTotal(selectExpenses(state.expenses, state.filters))
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
