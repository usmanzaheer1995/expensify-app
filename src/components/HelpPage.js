import React from 'react';
const HelpPage = () => <div>Help page Component.</div>;

export default HelpPage;
