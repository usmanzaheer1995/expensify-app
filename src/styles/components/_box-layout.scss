.box-layout {
    background: url('/images/bg.jpg');
    background-size: cover;
    height: 100vh;
    width: 100vw;

    display: flex;
    justify-content: center;
    align-items: center;

    &__box {
        background: fade-out($color: white, $amount: .15);
        border-radius: 3px;
        padding: $l-size $m-size;
        text-align: center;
        width: 25rem;
    }

    &__title {
        margin: 0 0 $m-size 0;
        line-height: 1;
    }

    &__button {
        width: 75%;
        background: #3F7FBF;
        color: fade-out($color: white, $amount: .15);;
        font-size: 18px;
        padding: 10px;
    }
}