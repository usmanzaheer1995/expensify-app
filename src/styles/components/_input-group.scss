.input-group {
    align-items: center;
    display: flex;
    margin-bottom: $l-size;

    &__item {
        margin-bottom: $s-size;
        @media only screen and (min-width: $desktop-breakpoint) {
            margin-right: $s-size;
            margin: 0 $s-size 0 0;
        }
    }

    @media only screen and (max-width: $desktop-breakpoint) {
        flex-direction: column;
        margin-bottom: $m-size;
    }
}