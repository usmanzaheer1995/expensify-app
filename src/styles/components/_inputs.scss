.text-input {
    border: 1px solid #cacccd;
    height: 5rem;
    font-size: $font-size-lg;
    font-weight: 300;
    padding: $s-size;
}

.textarea {
    @extend .text-input;
    height: 10rem;
}