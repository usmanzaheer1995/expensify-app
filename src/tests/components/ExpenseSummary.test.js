import React from "react";
import { shallow } from "enzyme";

import { ExpenseSummary } from "../../components/ExpenseSummary";
import expenses from "../fixtures/expenses";

test("should correctly render ExpensesSummary with one expense", () => {
  const wrapper = shallow(
    <ExpenseSummary expenses={[expenses[0]]} expensesTotal={35} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should correctly render ExpensesSummary with multiple expenses", () => {
  const wrapper = shallow(
    <ExpenseSummary expenses={expenses} expensesTotal={123456789} />
  );
  expect(wrapper).toMatchSnapshot();
});
