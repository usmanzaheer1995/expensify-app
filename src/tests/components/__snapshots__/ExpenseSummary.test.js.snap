// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        three
      </span>
       
      expenses
       totaling
       
      <span>
        Rs 123,456,789
      </span>
      .
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        one
      </span>
       
      expense
       totaling
       
      <span>
        Rs 35
      </span>
      .
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
